Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,1109 @@
+    case .clientIdleServerIdle, .clientOpenServerIdle, .clientClosedServerIdle:
+      throw self.assertionFailureAndCreateRPCErrorOnInternalError("Server can't send status if idle.")
+    case .clientOpenServerClosed, .clientClosedServerClosed:
+      throw self.assertionFailureAndCreateRPCErrorOnInternalError("Server can't send anything if closed.")
+    }
+  }
+  
+  private mutating func serverReceive(
+    metadata: HPACKHeaders,
+    endStream: Bool,
+    configuration: GRPCStreamStateMachineConfiguration.ServerConfiguration
+  ) throws -> OnMetadataReceived {
+    if endStream, case .clientIdleServerIdle = self.state {
+      throw self.assertionFailureAndCreateRPCErrorOnInternalError(
+          """
+          Client should have opened before ending the stream: \
+          stream shouldn't have been closed when sending initial metadata.
+          """
+      )
+    }
+    
+    switch self.state {
+    case .clientIdleServerIdle(let state):
+      guard metadata.contentType != nil else {
+        throw RPCError(code: .invalidArgument, message: "Invalid or empty \(GRPCHTTP2Keys.contentType.rawValue).")
+      }
+      
+      guard metadata.path != nil else {
+        throw RPCError(code: .unimplemented, message: "No \(GRPCHTTP2Keys.path.rawValue) header has been set.")
+      }
+      // TODO: Should we verify the RPCRouter can handle this endpoint here,
+      // or should we verify that in the handler?
+      
+      var outboundEncoding: CompressionAlgorithm? = nil
+      if let clientAdvertisedEncodings = metadata.acceptedEncodings {
+        for clientAcceptedEncoding in clientAdvertisedEncodings where configuration.acceptedEncodings.contains(where: { $0 == clientAcceptedEncoding }) {
+          // Found the preferred encoding: use it to compress responses.
+          outboundEncoding = clientAcceptedEncoding
+          break
+        }
+      }
+      
+      var inboundEncoding: CompressionAlgorithm? = nil
+      let encodingValues = metadata[GRPCHTTP2Keys.encoding.rawValue]
+      var encodingValuesIterator = encodingValues.makeIterator()
+      if let rawEncoding = encodingValuesIterator.next() {
+        guard encodingValuesIterator.next() == nil else {
+          throw RPCError(
+            code: .invalidArgument,
+            message: "\(GRPCHTTP2Keys.encoding) must contain no more than one value."
+          )
+        }
+        
+        guard let clientEncoding = CompressionAlgorithm(rawValue: rawEncoding), configuration.acceptedEncodings.contains(where: { $0 == clientEncoding }) else {
+          if configuration.acceptedEncodings.isEmpty {
+            throw RPCError(
+              code: .unimplemented,
+              message: "Compression is not supported"
+            )
+          } else {
+            let status = Status(
+              code: .unimplemented,
+              message: "\(rawEncoding) compression is not supported; supported algorithms are listed in grpc-accept-encoding"
+            )
```

Comment:
We'll need to encode these directly into the headers since that'll be what's returned to the client. It also means we need to form a trailers-only response and move into the server closed state.